skimage/restoration/non_local_means.py

```diff
@@ -0,0 +1,125 @@
+    provided that the *patches* centered on the other pixels are similar enough
+    to the patch centered on the pixel of interest.
+
+    In the original version of the algorithm [1]_, corresponding to
+    ``fast=False``, the computational complexity is
+
+    image.size * patch_size ** image.ndim * patch_distance ** image.ndim
+
+    Hence, changing the size of patches or their maximal distance has a
+    strong effect on computing times, especially for 3-D images.
+
+    However, the default behavior corresponds to ``fast=True``, for which
+    another version of non-local means [2]_ is used, corresponding to a
+    complexity of
+
+    image.size * patch_distance ** image.ndim
+
+    The computing time depends only weakly on the patch size, thanks to the
+    computation of the integral of patches distances for a given shift, that
+    reduces the number of operations [1]_. Therefore, this algorithm executes
+    faster than `nl_means_denoising`, at the expense of using twice as much
+    memory.
+
+    Compared to the classic non-local means algorithm implemented in
+    `nl_means_denoising`, all pixels of a patch contribute to the distance to
+    another patch with the same weight, no matter their distance to the center
+    of the patch. This coarser computation of the distance can result in a
+    slightly poorer denoising performance. Moreover, for small images (images
+    with a linear size that is only a few times the patch size), the classic
+    algorithm can be faster due to boundary effects.
+
+    The image is padded using the `reflect` mode of `skimage.util.pad`
+    before denoising.
+
+    References
+    ----------
+    .. [1] Buades, A., Coll, B., & Morel, J. M. (2005, June). A non-local
+        algorithm for image denoising. In CVPR 2005, Vol. 2, pp. 60-65, IEEE.
+
+    .. [2] Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
+           Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
+
+    Examples
+    --------
+    >>> a = np.zeros((40, 40))
+    >>> a[10:-10, 10:-10] = 1.
+    >>> a += 0.3*np.random.randn(*a.shape)
+    >>> denoised_a = nl_means_denoising(a, 7, 5, 0.1)
+    """
+    if image.ndim == 2:
+        if fast_mode:
+            return np.array(_fast_nl_means_denoising_2d(image, s=patch_size,
+                                            d=patch_distance, h=h))
+        else:
+            return np.array(_nl_means_denoising_2d(image, s=patch_size,
+                                d=patch_distance, h=h))
+    if image.ndim == 3 and image.shape[-1] > 4:  # only grayscale
+        if fast_mode:
+            return np.array(_fast_nl_means_denoising_3d(image, s=patch_size,
+                                              d=patch_distance, h=h))
+        else:
+            return np.array(_nl_means_denoising_3d(image, patch_size,
+                                patch_distance, h))
+    if image.ndim == 3 and image.shape[-1] == 3:  # 2-D color (RGB) images
```

Comment:
I'm quite happy with this, but when we discussed this "magic" in the context of my SLIC PR, the conclusion was that it's better to have a user flag, such as `rgb=True`, that determines whether an image is interpreted as 3D or 2D+c. The rationale was provided by @JDWarner: if someone has a little test 3D image with, say, `image.shape[-1] == 5`, on which they are developing their whole code, they will do the NLmeans denoising and be happy with their code. Then when they deploy it on many test images, they might come across a 3D image with `image.shape[-1] == 3`, and the code will do something completely different. In the case of a kwarg, users are protected from such a scenario.

The flag name is `multichannel` for SLIC and although it only makes much sense with RGB here, I would suggest we maintain the same flag name across all functions.

Also, right now, 3D images with exactly 4 columns in the last dimension just fail. =P